File: day09/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()
	scanner := bufio.NewScanner(file)

	r := NewTargetReader(scanner)

	target, err := r.Scan()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Part 1:", target)

	sum := 0
	buffer := []int{}
	for sum != target {
		if sum < target {
			k, err := r.Next()
			if err != nil {
				log.Fatal(err)
			}
			sum += k
			buffer = append(buffer, k)
		} else {
			if len(buffer) == 0 {
				break
			}
			sum -= buffer[0]
			buffer = buffer[1:]
		}
	}
	if sum == target {
		min, max, ok := minMax(buffer)
		if ok {
			fmt.Println("Part 2:", min+max)
		}
	}
}

type (
	TargetReader struct {
		buf     []int
		scanner *bufio.Scanner
	}
)

func NewTargetReader(scanner *bufio.Scanner) *TargetReader {
	return &TargetReader{
		buf:     []int{},
		scanner: scanner,
	}
}

func (r *TargetReader) readLine() (int, error) {
	if !r.scanner.Scan() {
		if err := r.scanner.Err(); err != nil {
			return 0, err
		}
		return 0, fmt.Errorf("End of file")
	}
	num, err := strconv.Atoi(r.scanner.Text())
	if err != nil {
		return 0, err
	}
	return num, nil
}

func (r *TargetReader) Scan() (int, error) {
	preamble := 25
	b := []int{}
	for preamble > 0 {
		k, err := r.readLine()
		if err != nil {
			return 0, err
		}
		b = append(b, k)
		preamble--
	}

	idx := 0
	for {
		k, err := r.readLine()
		if err != nil {
			return 0, err
		}
		b = append(b, k)
		if !hasParts(k, b[idx:idx+25]) {
			break
		}
		idx++
	}

	r.buf = b
	return b[len(b)-1], nil
}

func (r *TargetReader) Next() (int, error) {
	if len(r.buf) > 0 {
		k := r.buf[0]
		r.buf = r.buf[1:]
		return k, nil
	}
	return r.readLine()
}

func hasParts(a int, buffer []int) bool {
	for i := 0; i < len(buffer); i++ {
		for j := i + 1; j < len(buffer); j++ {
			if buffer[i]+buffer[j] == a {
				return true
			}
		}
	}
	return false
}

func minMax(b []int) (int, int, bool) {
	if len(b) == 0 {
		return 0, 0, false
	}
	min := b[0]
	max := b[0]
	for _, i := range b {
		if i < min {
			min = i
		}
		if i > max {
			max = i
		}
	}
	return min, max, true
}

File: day09/src/main.rs
use std::collections::VecDeque;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

struct TargetReader {
    buf: Box<dyn Iterator<Item = i32>>,
    iter: Box<dyn Iterator<Item = std::io::Result<String>>>,
}

impl TargetReader {
    fn new(iter: Box<dyn Iterator<Item = std::io::Result<String>>>) -> TargetReader {
        TargetReader {
            buf: Box::new(Vec::new().into_iter()),
            iter,
        }
    }

    fn read_line(&mut self) -> Result<i32, BErr> {
        let line = self.iter.next().ok_or("EOF")??;
        let num = line.parse::<i32>()?;
        Ok(num)
    }

    fn scan(&mut self) -> Result<i32, BErr> {
        let mut b = Vec::new();
        for _ in 0..25 {
            let k = self.read_line()?;
            b.push(k);
        }

        let mut idx = 0;
        loop {
            let k = self.read_line()?;
            b.push(k);
            if !has_parts(k, &b[idx..idx + 25]) {
                break;
            }
            idx += 1;
        }
        let last = b[b.len() - 1];
        self.buf = Box::new(b.into_iter());
        Ok(last)
    }

    fn next_int(&mut self) -> Result<i32, BErr> {
        if let Some(k) = self.buf.next() {
            return Ok(k);
        }
        self.read_line()
    }
}

fn has_parts(a: i32, b: &[i32]) -> bool {
    for i in 0..b.len() {
        for j in i + 1..b.len() {
            if b[i] + b[j] == a {
                return true;
            }
        }
    }
    false
}

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut r = TargetReader::new(Box::new(reader.lines()));

    let target = r.scan()?;
    println!("Part 1: {}", target);

    let mut sum = 0;
    let mut buf = VecDeque::new();
    while sum != target {
        if sum < target {
            let k = r.next_int()?;
            sum += k;
            buf.push_back(k);
        } else {
            if buf.len() == 0 {
                break;
            }
            let first = buf[0];
            sum -= first;
            buf.pop_front();
        }
    }
    if sum == target {
        buf.make_contiguous();
        let (slice, _) = buf.as_slices();
        let (min, max) = min_max(slice)?;
        println!("Part 2: {}", min + max);
    }
    Ok(())
}

fn min_max(b: &[i32]) -> Result<(i32, i32), BErr> {
    if b.len() == 0 {
        return Err("No elements".into());
    }
    let mut min = *b.first().ok_or("No elements")?;
    let mut max = min;
    for &i in b {
        if i < min {
            min = i;
        }
        if i > max {
            max = i;
        }
    }
    Ok((min, max))
}
